### ISSUE TITLE
EmptyHeader terminating the stream with "Entity stream truncation"

### ISSUE THREAD

**Issue by supercharger** [akka/akka-http#2786]
When used for HTTP transfer chunked encoding (server is Akka as well), We are seeing following exception intermittently which is very difficult to reproduce it.  It looks like this exception is originating at this  [HttpMessageParser.scala#L143 ](https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala#L143).  What does EmptyHeader mean, Is there a reason to terminate the stream/throw exception instead of ignoring it.

One thing that needs to be fixed for certain is that this case **doesn't have any logging to indicate what went wrong** before throwing the exception, so it needs to be added. We've enabled Debug logging, but the logging doesn't indicate what went wrong.

```
Caused by: 
akka.http.scaladsl.model.EntityStreamException: Entity stream truncation
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$2.applyOrElse(HttpMessageParser.scala:348)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$2.applyOrElse(HttpMessageParser.scala:346)
        at akka.stream.impl.fusing.Collect$$anon$6.$anonfun$wrappedPf$1(Ops.scala:227)
        at akka.stream.impl.fusing.SupervisedGraphStageLogic.withSupervision(Ops.scala:186)
        at akka.stream.impl.fusing.Collect$$anon$6.onPush(Ops.scala:229)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:523)
        at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:480)
```
Thanks for your time in looking in to this, appreciate it. I can raise a PR if more details on EmptyHeader can be provided.

Debug logs ( which doesn't help ) :
```
DEBUG] [10/20/2019 16:30:20.425] [main] [AkkaSSLConfig(akka://ImportUserPreferences)] Initializing AkkaSSLConfig extension...
[DEBUG] [10/20/2019 16:30:20.428] [main] [AkkaSSLConfig(akka://ImportUserPreferences)] buildHostnameVerifier: created hostname verifier: com.typesafe.sslconfig.ssl.NoopHostnameVerifier@4e3
01795
[DEBUG] [10/20/2019 16:30:21.408] [ImportUserPreferences-akka.actor.default-dispatcher-3] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a] Executing [WorkerForCommand(Connect(someserver:8080,None,List(),Some(10 seconds),true),Actor[akka://ImportUserPreferences/system/StreamSupervisor-0/$$a#724857595],akka.io.TcpManager$$anonfun$receive$1$$La
mbda$426/660205305@7f2157da)]
[DEBUG] [10/20/2019 16:30:21.418] [ImportUserPreferences-akka.actor.default-dispatcher-4] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a] Scheduling Registering channel java.nio.
channels.SocketChannel[unconnected] with initialOps 0
[DEBUG] [10/20/2019 16:30:21.418] [ImportUserPreferences-akka.io.pinned-dispatcher-5] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a] Registering channel java.nio.channels.Socket
Channel[unconnected] with initialOps 0
[DEBUG] [10/20/2019 16:30:21.421] [ImportUserPreferences-akka.actor.default-dispatcher-4] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Resolving someserver before connecting
[DEBUG] [10/20/2019 16:30:21.446] [ImportUserPreferences-akka.actor.default-dispatcher-2] [akka://ImportUserPreferences/system/IO-DNS] Resolution request for someserver from Actor[akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0#2021369182]
[DEBUG] [10/20/2019 16:30:21.463] [ImportUserPreferences-akka.actor.default-dispatcher-2] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Attempting connection to [someserver:8080]
[DEBUG] [10/20/2019 16:30:21.464] [ImportUserPreferences-akka.io.pinned-dispatcher-5] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a] Enabling 8 on sun.nio.ch.SelectionKeyImpl@47
f50277
[DEBUG] [10/20/2019 16:30:21.473] [ImportUserPreferences-akka.actor.default-dispatcher-2] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Connection established to [someserver:8080]
[DEBUG] [10/20/2019 16:30:21.481] [ImportUserPreferences-akka.actor.default-dispatcher-11] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] [Actor[akka://ImportUserPreferences/s
ystem/StreamSupervisor-0/$$a#724857595]] registered as connection handler
[DEBUG] [10/20/2019 16:30:21.488] [ImportUserPreferences-akka.io.pinned-dispatcher-5] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a] Enabling 1 on sun.nio.ch.SelectionKeyImpl@47
f50277
[DEBUG] [10/20/2019 16:30:21.505] [ImportUserPreferences-akka.actor.default-dispatcher-11] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Wrote [158] bytes to channel
[DEBUG] [10/20/2019 16:30:21.736] [ImportUserPreferences-akka.actor.default-dispatcher-3] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Read [161] bytes.
[DEBUG] [10/20/2019 16:30:21.770] [ImportUserPreferences-akka.io.pinned-dispatcher-5] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a] Enabling 1 on sun.nio.ch.SelectionKeyImpl@47
f50277
[DEBUG] [10/20/2019 16:30:31.543] [ImportUserPreferences-akka.actor.default-dispatcher-11] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Read [-1] bytes.
[DEBUG] [10/20/2019 16:30:31.544] [ImportUserPreferences-akka.actor.default-dispatcher-11] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Read returned end-of-stream, our side
 not yet closed
[DEBUG] [10/20/2019 16:30:31.552] [ImportUserPreferences-akka.actor.default-dispatcher-4] [akka://ImportUserPreferences/system/IO-TCP/selectors/$a/0] Got Close command, closing connection
```

**Comment by jrudolph** [comment]
Hi @supercharger, thanks for the report. Does that error happen on the client-side?

The error message means that the connection was closed / aborted when still more data was expected. The reason could be that either the server has a bug, or the server processing fails while already started to send out a response in which case the only way forward is to abort the connection. Or the connection was completed for another reason. Unfortunately, there's no way to communicate a reason for a premature closing of the connection from server to client in that stage. That's also the reason that the client cannot show any better error message.

You could enable `akka.http.client.log-unencrypted-network-bytes = 1000` to debug log what's being received on the wire. What you will probably see there, is that for chunked encoding, the final `0` length chunk is missing.

> When used for HTTP transfer chunked encoding (server is Akka as well), We are seeing following exception intermittently which is very difficult to reproduce it. It looks like this exception is originating at this [HttpMessageParser.scala#L143 ](https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala#L143) [](/akka/akka-http/blob/HEAD@{2019-10-24T16:04:46Z}/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala#L143)

The `EmptyHeader` is not significant here. In the parser it means that it has completed reading headers and is now expecting the body of the request. In that moment it sets a completion handler that will fail the response if the entity cannot be read fully. That completion handler is later reset when the response was read successfully.

**Comment by jrudolph** [comment]
I suggest to improve the error message for that case slightly in https://github.com/akka/akka-http/pull/2789. Would that have helped you?

**Comment by supercharger** [comment]
Thanks for your help, #2789  definitely helps. I will try with `akka.http.client.log-unencrypted-network-bytes`

One more thing I want to point is that,  [Wrapping the source with ](https://doc.akka.io/docs/akka/current/stream/operators/RestartSource/withBackoff.html)**`RestartSource.withBackoff`** doesn't work for Entity Stream Exception, while it works for initial connection level issues.  Also, Using Supervisor to catch EntityStreamException and [restarting](https://doc.akka.io/japi/akka/current/akka/actor/SupervisorStrategy.html#restart()) seems to loose certain streaming messages. 

What's the best way to retry when we facing with EntityStreamException  ( as above doesn't work) ?

**Comment by supercharger** [comment]
@jrudolph Just want to check with you on why `RestartSource` doesn't work, I provided more details on the earlier/above comment.

**Comment by jrudolph** [comment]
I can imagine that using RestartSource would be difficult in regard. What would you like the HTTP behavior to be on the wire exactly?

In general, I would recommend using `Http().singleRequest` and then adding a `watchTermination` hook to the response entity source to deal with entity stream problems before handling the request. However, a general problem when using streaming is that you might have already started processing the response entity, executing some side-effects based on the incoming data and then it fails finally. Then you might have to do some cleanup before retrying the request.

**Comment by supercharger** [comment]
> What would you like the HTTP behavior to be on the wire exactly?

We generally expect it to be retried for idempotent HTTP methods. `RestartSource` documentation ( along with examples) clearly points out that it can retried while in the middle of the stream. If there are some side effects with consuming data multiple times, `RestartSource` might not be best fit as documentation.

From documentation of `watchTermination`, I see it allows to handle the stream with `done` or `fail`.  But with `watchTermination`, it seems to became hacky code with recursion.  I will try with multiple ( retry count)  `CompletionStage`  chaining/joining on exceptionally.

Thanks for you help.

**Comment by jrudolph** [comment]
This seems to have become more of a usage question. If it's still a problem could you take it to the discussion forums at https://discuss.akka.io? Thanks!

**Comment by therockstorm** [comment]
We were streaming files from S3 through a file service and on to a second service using Akka HTTP streams. What fixed it for us was this config in both services,

```
akka {
  http {
    client {
      # Large files with expensive processing back-pressure, causing stream to close before processed
      idle-timeout = infinite # Default 60s

      parsing {
        max-content-length = infinite # Default infinite, but without this it fails on large files
      }
    }

    host-connection-pool {
      # Without both of these and `client.idle-timeout` above, saw this with large files,
      # `Entity stream truncation. The HTTP parser was receiving an entity when the underlying connection was closed unexpectedly.`
      idle-timeout = infinite # Default 30s
      response-entity-subscription-timeout = 30s # Default 1s
    }
  }
}
```

**Comment by michalstutzmann** [comment]
@therockstorm have you managed to solve the problem eventually?

**Comment by therockstorm** [comment]
We did not. We ended up removing the feature in part because of this issue.

**Comment by yahor-ivanou** [comment]
Facing the same issue with closed connection while using akka http client and processing response with huge file(100MB+) using `Http().singleRequest()`

I am making 2 concurrent requests to download **the same file from the same host** (e.g. http://ipv4.download.thinkbroadband.com/100MB.zip) but having for the processing of the response only one thread, so that `response.dataBytes` from the responses are being processed consequently.

The processing of the first response is always successful, but the second one is always failing with `ErrorInfo(Entity stream truncation. The HTTP parser was receiving an entity when the underlying connection was closed unexpectedly., (details omitted), )`
I have tried all possible timeouts(client and host-connection-pool) to prevent the connection from being closed, but no go.

When I try to make two requests, but each request goes to a different host, then everything works fine. Both responses are being processed successfully.
Having 2 threads also fixes the problem.
The setting `host-connection-pool.max-connections = 1`(default is 1024) also helps. But, I think that it is an ugly workaround.

Any ideas how to fix closed connection without workarounds?
